pkg/identityserver/store/eui_block_store.go

```diff
@@ -0,0 +1,161 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"context"
+	"fmt"
+	"runtime/trace"
+
+	"github.com/jinzhu/gorm"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/types"
+)
+
+func GetEUIStore(db *gorm.DB) EUIStore {
+	return &euiStore{store: newStore(db)}
+}
+
+type euiStore struct {
+	*store
+}
+
+func getMaxAddress(addressBlock types.EUI64Prefix) *EUI64 {
+	var maxAddress types.EUI64
+	maxAddress64 := addressBlock.EUI64.MarshalNumber() | (^(uint64(0)) >> (64 - addressBlock.Length))
+	maxAddress.UnmarshalNumber(maxAddress64)
+	return eui(&maxAddress)
+}
+
+var (
+	errMaxGlobalEUILimitReached = errors.DefineInternal("global_eui_limit_reached", "global eui limit from address block reached")
+	errAppDevEUILimitReached    = errors.DefineInternal("application_dev_eui_limit_reached", "application issued DevEUI limit ({dev_eui_limit}) reached")
+)
+
+func (s *euiStore) incrementApplicationDevEUICounter(ctx context.Context, ids *ttnpb.ApplicationIdentifiers, appDevEUILimit int) error {
+	var appModel Application
+	// Check if application exists.
+	query := s.query(WithoutSoftDeleted(ctx), Application{}, withApplicationID(ids.GetApplicationId()))
+	if err := query.First(&appModel).Error; err != nil {
+		if gorm.IsRecordNotFoundError(err) {
+			return errNotFoundForID(ids)
+		}
+		return err
+	}
+	// If app dev eui limit not configured, skip this step.
+	if appDevEUILimit == 0 {
+		return nil
+	}
+	// Atomically check and increment dev eui counter if its below threshold.
+	result := s.query(ctx, Application{}).
+		Where(Application{ApplicationID: appModel.ApplicationID}).
+		Where(fmt.Sprintf(`"applications"."dev_eui_counter" < '%d'`, appDevEUILimit)).
+		Update("dev_eui_counter", gorm.Expr("dev_eui_counter + ?", 1))
+	// If application dev eui counter not updated, the limit was reached.
+	if result.RowsAffected == 0 {
+		return errAppDevEUILimitReached.WithAttributes("dev_eui_limit", fmt.Sprint(appDevEUILimit))
+	}
+	if result.Error != nil {
+		return result.Error
+	}
+	return nil
+}
+
+func (s *euiStore) issueDevEUIAddressFromBlock(ctx context.Context) (*types.EUI64, error) {
+	for {
+		// Check if DevEUI block limit exists.
+		query := s.query(ctx, EUIBlock{}).
+			Select(`"eui_blocks"."current_address" AS "eui", "end_devices"."id" AS "existing_id"`).
+			Joins(`LEFT JOIN "end_devices" ON "end_devices"."dev_eui" = "eui_blocks"."current_address"`).
+			Where(`"eui_blocks"."type" = 'dev_eui' AND "eui_blocks"."current_address" <> "eui_blocks"."end_address"`)
+
+		var result struct {
+			Eui        EUI64
+			ExistingId string
+		}
+		if err := query.Scan(&result).Error; err != nil {
+			if gorm.IsRecordNotFoundError(err) {
+				return nil, errMaxGlobalEUILimitReached
+			}
+			return nil, err
+		}
+		if err := s.query(ctx, EUIBlock{}).
```

Comment:
This is not atomic and can race between the Identity Servers that are just starting up, I think.